File: src/data/repository/PersonalRecordRepository.ts
/**
 * Personal Record Repository - Manages personal records (PRs)
 * Migrated from Android Kotlin PersonalRecordRepository to React Native TypeScript
 */

import * as PersonalRecordDao from '../local/daos/personalRecordDao';
import { PersonalRecordEntity } from '../local/entities';
import { PersonalRecord } from '../../domain/models/Models';

/**
 * Personal Record Repository interface
 */
export interface IPersonalRecordRepository {
  getLatestPR(exerciseId: string, workoutMode: string): Promise<PersonalRecord | null>;
  getPRsForExercise(exerciseId: string): Promise<PersonalRecord[]>;
  getBestPR(exerciseId: string): Promise<PersonalRecord | null>;
  getAllPRs(): Promise<PersonalRecord[]>;
  getAllPRsGrouped(): Promise<PersonalRecord[]>;
  updatePRIfBetter(
    exerciseId: string,
    weightPerCableKg: number,
    reps: number,
    workoutMode: string,
    timestamp: number
  ): Promise<boolean>;
}

/**
 * Personal Record Repository implementation
 */
class PersonalRecordRepositoryImpl implements IPersonalRecordRepository {
  constructor() {}

  /**
   * Get the latest PR for an exercise in a specific workout mode
   */
  async getLatestPR(exerciseId: string, workoutMode: string): Promise<PersonalRecord | null> {
    try {
      const entity = await PersonalRecordDao.getLatestPR(exerciseId, workoutMode);
      return entity ? this.entityToPersonalRecord(entity) : null;
    } catch (error) {
      console.error(`[PersonalRecordRepository] Failed to get PR for exercise ${exerciseId}:`, error);
      return null;
    }
  }

  /**
   * Get all PRs for an exercise across all workout modes
   */
  async getPRsForExercise(exerciseId: string): Promise<PersonalRecord[]> {
    try {
      const entities = await PersonalRecordDao.getPRsForExercise(exerciseId);
      return entities.map(this.entityToPersonalRecord);
    } catch (error) {
      console.error(`[PersonalRecordRepository] Failed to get PRs for exercise ${exerciseId}:`, error);
      return [];
    }
  }

  /**
   * Get the best PR for an exercise across all modes
   */
  async getBestPR(exerciseId: string): Promise<PersonalRecord | null> {
    try {
      const entity = await PersonalRecordDao.getBestPR(exerciseId);
      return entity ? this.entityToPersonalRecord(entity) : null;
    } catch (error) {
      console.error(`[PersonalRecordRepository] Failed to get best PR for exercise ${exerciseId}:`, error);
      return null;
    }
  }

  /**
   * Get all personal records
   */
  async getAllPRs(): Promise<PersonalRecord[]> {
    try {
      const entities = await PersonalRecordDao.getAllPRs();
      return entities.map(this.entityToPersonalRecord);
    } catch (error) {
      console.error('[PersonalRecordRepository] Failed to get all PRs:', error);
      return [];
    }
  }

  /**
   * Get all personal records grouped by exercise (for analytics)
   */
  async getAllPRsGrouped(): Promise<PersonalRecord[]> {
    try {
      const entities = await PersonalRecordDao.getAllPRsGrouped();
      return entities.map(this.entityToPersonalRecord);
    } catch (error) {
      console.error('[PersonalRecordRepository] Failed to get grouped PRs:', error);
      return [];
    }
  }

  /**
   * Update PR if the new performance is better
   * Returns true if a new PR was set, false otherwise
   */
  async updatePRIfBetter(
    exerciseId: string,
    weightPerCableKg: number,
    reps: number,
    workoutMode: string,
    timestamp: number
  ): Promise<boolean> {
    try {
      const isNewPR = await PersonalRecordDao.updatePRIfBetter(
        exerciseId,
        weightPerCableKg,
        reps,
        workoutMode,
        timestamp
      );

      if (isNewPR) {
        console.log(
          `[PersonalRecordRepository] New PR set for exercise ${exerciseId}: ${weightPerCableKg}kg x ${reps} reps (${workoutMode})`
        );
      }

      return isNewPR;
    } catch (error) {
      console.error(`[PersonalRecordRepository] Failed to update PR for exercise ${exerciseId}:`, error);
      return false;
    }
  }

  // ========== Helper Methods ==========

  /**
   * Convert PersonalRecordEntity to PersonalRecord
   */
  private entityToPersonalRecord(entity: PersonalRecordEntity): PersonalRecord {
    return {
      id: entity.id,
      exerciseId: entity.exerciseId,
      weightPerCableKg: entity.weightPerCableKg,
      reps: entity.reps,
      timestamp: entity.timestamp,
      workoutMode: entity.workoutMode,
    };
  }
}

// Export singleton instance
let personalRecordRepositoryInstance: PersonalRecordRepositoryImpl | null = null;

export const getPersonalRecordRepository = (): IPersonalRecordRepository => {
  if (!personalRecordRepositoryInstance) {
    personalRecordRepositoryInstance = new PersonalRecordRepositoryImpl();
  }
  return personalRecordRepositoryInstance;
};

export const resetPersonalRecordRepository = (): void => {
  personalRecordRepositoryInstance = null;
};

// Export types
export type { IPersonalRecordRepository };
export { PersonalRecordRepositoryImpl };

File: deprecated/app/src/main/java/com/example/vitruvianredux/data/repository/PersonalRecordRepository.kt
package com.example.vitruvianredux.data.repository

import com.example.vitruvianredux.data.local.PersonalRecordDao
import com.example.vitruvianredux.data.local.PersonalRecordEntity
import com.example.vitruvianredux.domain.model.PersonalRecord
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import timber.log.Timber
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Repository for managing personal records (PRs)
 */
@Singleton
class PersonalRecordRepository @Inject constructor(
    private val personalRecordDao: PersonalRecordDao
) {

    /**
     * Get the latest PR for an exercise in a specific workout mode
     */
    suspend fun getLatestPR(exerciseId: String, workoutMode: String): PersonalRecord? {
        return try {
            personalRecordDao.getLatestPR(exerciseId, workoutMode)?.toPersonalRecord()
        } catch (e: Exception) {
            Timber.e(e, "Failed to get PR for exercise $exerciseId")
            null
        }
    }

    /**
     * Get all PRs for an exercise across all workout modes
     */
    fun getPRsForExercise(exerciseId: String): Flow<List<PersonalRecord>> {
        return personalRecordDao.getPRsForExercise(exerciseId).map { entities ->
            entities.map { it.toPersonalRecord() }
        }
    }

    /**
     * Get the best PR for an exercise across all modes
     */
    suspend fun getBestPR(exerciseId: String): PersonalRecord? {
        return try {
            personalRecordDao.getBestPR(exerciseId)?.toPersonalRecord()
        } catch (e: Exception) {
            Timber.e(e, "Failed to get best PR for exercise $exerciseId")
            null
        }
    }

    /**
     * Get all personal records
     */
    fun getAllPRs(): Flow<List<PersonalRecord>> {
        return personalRecordDao.getAllPRs().map { entities ->
            entities.map { it.toPersonalRecord() }
        }
    }

    /**
     * Get all personal records grouped by exercise (for analytics)
     */
    fun getAllPRsGrouped(): Flow<List<PersonalRecord>> {
        return personalRecordDao.getAllPRsGrouped().map { entities ->
            entities.map { it.toPersonalRecord() }
        }
    }

    /**
     * Update PR if the new performance is better
     * Returns Result.success(true) if a new PR was set, Result.success(false) otherwise
     */
    suspend fun updatePRIfBetter(
        exerciseId: String,
        weightPerCableKg: Float,
        reps: Int,
        workoutMode: String,
        timestamp: Long
    ): Result<Boolean> {
        return try {
            val isNewPR = personalRecordDao.updatePRIfBetter(
                exerciseId = exerciseId,
                weightPerCableKg = weightPerCableKg,
                reps = reps,
                workoutMode = workoutMode,
                timestamp = timestamp
            )
            Result.success(isNewPR)
        } catch (e: Exception) {
            Timber.e(e, "Failed to update PR for exercise $exerciseId")
            Result.failure(e)
        }
    }
}

// Extension functions for mapping between entities and domain models
private fun PersonalRecordEntity.toPersonalRecord() = PersonalRecord(
    id = id,
    exerciseId = exerciseId,
    weightPerCableKg = weightPerCableKg,
    reps = reps,
    timestamp = timestamp,
    workoutMode = workoutMode
)
